// @taquito/rpc/dist/lib/opkind.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpKind = void 0;
var OpKind;
(function (OpKind) {
    OpKind["ORIGINATION"] = "origination";
    OpKind["DELEGATION"] = "delegation";
    OpKind["REVEAL"] = "reveal";
    OpKind["TRANSACTION"] = "transaction";
    OpKind["ACTIVATION"] = "activate_account";
    OpKind["ATTESTATION"] = "attestation";
    OpKind["ENDORSEMENT"] = "endorsement";
    OpKind["PREATTESTATION"] = "preattestation";
    OpKind["PREENDORSEMENT"] = "preendorsement";
    OpKind["SET_DEPOSITS_LIMIT"] = "set_deposits_limit";
    OpKind["DOUBLE_PREATTESTATION_EVIDENCE"] = "double_preattestation_evidence";
    OpKind["DOUBLE_PREENDORSEMENT_EVIDENCE"] = "double_preendorsement_evidence";
    OpKind["ATTESTATION_WITH_SLOT"] = "attestation_with_slot";
    OpKind["ENDORSEMENT_WITH_SLOT"] = "endorsement_with_slot";
    OpKind["SEED_NONCE_REVELATION"] = "seed_nonce_revelation";
    OpKind["DOUBLE_ATTESTATION_EVIDENCE"] = "double_attestation_evidence";
    OpKind["DOUBLE_ENDORSEMENT_EVIDENCE"] = "double_endorsement_evidence";
    OpKind["DOUBLE_BAKING_EVIDENCE"] = "double_baking_evidence";
    OpKind["PROPOSALS"] = "proposals";
    OpKind["BALLOT"] = "ballot";
    OpKind["FAILING_NOOP"] = "failing_noop";
    OpKind["REGISTER_GLOBAL_CONSTANT"] = "register_global_constant";
    OpKind["TRANSFER_TICKET"] = "transfer_ticket";
    OpKind["INCREASE_PAID_STORAGE"] = "increase_paid_storage";
    OpKind["UPDATE_CONSENSUS_KEY"] = "update_consensus_key";
    OpKind["DRAIN_DELEGATE"] = "drain_delegate";
    OpKind["VDF_REVELATION"] = "vdf_revelation";
    OpKind["EVENT"] = "event";
    OpKind["TICKET_UPDATES"] = "ticket_updates";
    OpKind["SMART_ROLLUP_ORIGINATE"] = "smart_rollup_originate";
    OpKind["SMART_ROLLUP_ADD_MESSAGES"] = "smart_rollup_add_messages";
    OpKind["SMART_ROLLUP_EXECUTE_OUTBOX_MESSAGE"] = "smart_rollup_execute_outbox_message";
    OpKind["SMART_ROLLUP_PUBLISH"] = "smart_rollup_publish";
    OpKind["SMART_ROLLUP_CEMENT"] = "smart_rollup_cement";
    OpKind["SMART_ROLLUP_RECOVER_BOND"] = "smart_rollup_recover_bond";
    OpKind["SMART_ROLLUP_REFUTE"] = "smart_rollup_refute";
    OpKind["SMART_ROLLUP_TIMEOUT"] = "smart_rollup_timeout";
})(OpKind || (exports.OpKind = OpKind = {}));
